iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/IamIntrospectionEndpoint.java

```diff
@@ -0,0 +1,266 @@
+    this.accountService = accountService;
+    this.revocationService = revocationService;
+  }
+
+  @PostMapping(value = "/introspect", consumes = {MediaType.APPLICATION_FORM_URLENCODED_VALUE},
+      produces = {MediaType.APPLICATION_JSON_VALUE})
+  @PreAuthorize("hasRole('ROLE_CLIENT')")
+  public IntrospectionResponse introspect(
+      @RequestParam(value = OAuth2ParameterNames.TOKEN, required = true) String tokenValue,
+      @RequestParam(value = OAuth2ParameterNames.TOKEN_TYPE_HINT,
+          required = false) TokenTypeHint tokenType,
+      Authentication auth)
+      throws UnauthorizedClientException, ParseException, InvalidTokenException {
+
+    ClientDetailsEntity c = loadClient(auth);
+    validate(c, tokenValue);
+
+    if (tokenType == null) {
+      tokenType = valueFrom(tokenValue);
+    }
+
+    if (TokenTypeHint.REFRESH_TOKEN.equals(tokenType)) {
+      return introspectRefreshToken(tokenValue);
+    }
+    return introspectAccessToken(tokenValue);
+  }
+
+  private TokenTypeHint valueFrom(String tokenValue) throws ParseException {
+
+    JWT jwt = JWTParser.parse(tokenValue);
+    if (jwt instanceof PlainJWT) {
+      return TokenTypeHint.REFRESH_TOKEN;
+    }
+    if (jwt instanceof SignedJWT) {
+      return TokenTypeHint.ACCESS_TOKEN;
+    }
+    throw new InvalidTokenException("Expected a SignedJWT or PlainJWT object");
+  }
+
+  private void validate(ClientDetailsEntity c, String tokenValue)
+      throws UnauthorizedClientException, InvalidTokenException {
+
+    // check if client has been suspended
+    if (!c.isActive()) {
+      String errorMsg = String.format(SUSPENDED_CLIENT_ERROR, c.getClientId());
+      logger.error(errorMsg);
+      throw new UnauthorizedClientException(errorMsg);
+    }
+
+    // check if client is allowed to introspect tokens
+    if (!c.isAllowIntrospection()) {
+      String errorMsg = String.format(NOT_ALLOWED_CLIENT_ERROR, c.getClientId());
+      logger.error(errorMsg);
+      throw new UnauthorizedClientException(errorMsg);
+    }
+
+    // invalid null token to introspect
+    if (Strings.isNullOrEmpty(tokenValue)) {
+      String errorMsg = "Verify failed; token value is null";
+      logger.error(errorMsg);
+      throw new InvalidTokenException(errorMsg);
+    }
+  }
+
+  private IntrospectionResponse introspectRefreshToken(String token)
+      throws ParseException, InvalidTokenException {
+
+    OAuth2RefreshTokenEntity rt = tokenService.getRefreshToken(token);
+    if (rt.isExpired()) {
+      return IntrospectionResponse.inactive();
+    }
+    IntrospectionResponse.Builder builder = new IntrospectionResponse.Builder(true);
+    // base response: copy all claims
+    rt.getJwt().getJWTClaimsSet().getClaims().forEach(builder::addField);
+    // token type
+    builder.addField(OAuth2TokenIntrospectionClaimNames.TOKEN_TYPE, TokenTypeHint.REFRESH_TOKEN);
+    builder.addField(OAuth2TokenIntrospectionClaimNames.SUB,
+        rt.getJwt().getJWTClaimsSet().getSubject());
```

Comment:
Only for my personal taste, I would avoid to repeat `getJwt().getJWTClaimsSet()` or `rt.getClient()` several times but rather store them in a temp variable. Just for aesthetic (don't know about performance).

Reply:
Honestly... I thought a lot about doing this. Ok, I'll do it :)

Reply:
Done